src/tools/tools_execute.rs

```diff
@@ -232,9 +232,25 @@ async fn pp_run_tools(
                 "context_file".to_string(),
                 serde_json::to_string(&json_vec).unwrap()
             );
-            generated_other.push(message.clone());
+            let mut found_exact_same_message = false;
+            for original_msg in original_messages.iter().rev() {
+                if original_msg.role == "user" {
+                    break;
+                }
+                if original_msg.content == message.content {
```

Comment:
PartialEq for String has complexity of O(n), where n is length of a string. Assuming, context message are pretty darn big, it might cost a lot